PG16 compatibility: Resolve compilation issues

naisila opened the pull request:
This PR provides successful compilation against PG16Beta2. It does some necessary refactoring to prepare for full support of version 16, in PR #6952 .

We can directly merge this to main, without including the CI changes (I added just the build-16 test) and configure script.

Notes to reviewer:
- Please review commit per commit. It is simpler to follow
- There is one commit prefixed with "FIX THIS". Ignore it for now.

(the branch name is wrong :D should be pg16compile)

onderkalaci commented:
> There are two commits prefixed with "FIX THIS". Ignore them for now.

We normally try to avoid merging into main thing that are not yet ready.  Are you planning to fix those items before merging?

onderkalaci commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -737,13 +737,23 @@ AddEdgesForLockWaits(WaitGraph *waitGraph, PGPROC *waitingProc, PROCStack *remai
 	int conflictMask = lockMethodTable->conflictTab[waitingProc->waitLockMode];
 
 	/* iterate through the queue of processes holding the lock */
+#if PG_VERSION_NUM >= PG_VERSION_16
```
uf, this is very critical piece of code and now super hard to follow.

Is there a possible refactoring that could make this code easier to follow/review?

naisila replied:
Let me see what I can do: at the very very least I could add some comments.

tejeswarm replied:
I made an attempt but it's not a low hanging fruit. Since `AddEdgesForLockWaits` routine is small and it's the only place where `SHMQueueNext` is used, another option is to have two versions of the same routine, one using `SHMQueueNext` and the other using `dlist` model. This will look clean and easy to follow when reading the code.

tejeswarm commented on src/backend/distributed/metadata/metadata_cache.c:
```diff
@@ -83,7 +83,9 @@
 #include "utils/memutils.h"
 #include "utils/palloc.h"
 #include "utils/rel.h"
+#if PG_VERSION_NUM < PG_VERSION_16
```
Probably we don't need these macros, for lower versions, it will be redundant include.

naisila replied:
I don't prefer redundant includes to be honest. It is easy to remove the macros when we drop support in the future.

tejeswarm commented on src/backend/columnar/columnar_customscan.c:
```diff
@@ -541,7 +541,11 @@ ColumnarIndexScanAdditionalCost(PlannerInfo *root, RelOptInfo *rel,
+#if PG_VERSION_NUM >= PG_VERSION_16
+	double absIndexCorrelation = fabs(indexCorrelation);
+#else
 	double absIndexCorrelation = Abs(indexCorrelation);
```
What's the logic behind using fabs? Can we have compat macro for this, where Abs-> will turn to fabs?

tejeswarm commented on src/backend/distributed/commands/index.c:
```diff
@@ -558,7 +559,11 @@ ReindexStmtFindRelationOid(ReindexStmt *reindexStmt, bool missingOk)
 	{
 		relationId = RangeVarGetRelidExtended(reindexStmt->relation, lockmode,
 											  (missingOk) ? RVR_MISSING_OK : 0,
+#if PG_VERSION_NUM >= PG_VERSION_16
```
This is very hard to read, let me see what can be done.

tejeswarm commented on src/backend/distributed/commands/index.c:
```diff
@@ -558,7 +559,11 @@ ReindexStmtFindRelationOid(ReindexStmt *reindexStmt, bool missingOk)
 											  (missingOk) ? RVR_MISSING_OK : 0,
+#if PG_VERSION_NUM >= PG_VERSION_16
+											  RangeVarCallbackMaintainsTable, NULL);
+#else
```
This is very hard to read, let me see what can be done.

tejeswarm replied:
It's confusing why this remained intact `RangeVarCallbackOwnsRelation`

tejeswarm commented on src/backend/distributed/commands/multi_copy.c:
```diff
@@ -3187,26 +3190,47 @@ CheckCopyPermissions(CopyStmt *copyStatement)
 	                  is_from ? RowExclusiveLock : AccessShareLock);
 
-	range_table = CreateRangeTable(rel, required_access);
+	range_table = CreateRangeTable(rel);
```
@naisila I'm confused now, is this routine `CreateRangeTable` Citus or PG community? It appears to be from the Citus which has signature of ` CreateRangeTable(Relation rel)` ?

tejeswarm replied:
OK, I think I know what happened, you removed the second parameter in your commits, that explains :)

tejeswarm commented on src/backend/distributed/commands/multi_copy.c:
```diff
@@ -3187,26 +3190,47 @@ CheckCopyPermissions(CopyStmt *copyStatement)
+	/* create permission info for rte */
+	RTEPermissionInfo *perminfo = GetFilledPermissionInfo(rel->rd_id, rte->inh, required_access);
+#else
+	rte->requiredPerms = required_access;
```
Is it possible to move this into the above routine, and the only line that executes for pre-16. That way we can remove the version macro here? You can ignore, if you feel like it's easy here

naisila replied:
I couldn't quite understand what you mean. I wonder if you can write a suggestion box.

tejeswarm replied:
Something on these lines
![image](https://github.com/citusdata/citus/assets/44680808/43620add-9906-4d56-bdbf-c3f6a4df38c4)

tejeswarm replied:
If you use the method, you did for other routines in `pg_version_compat.h` , then we will have two versions of the same routine. If you think, it's cumbersome, please ignore my comment.

naisila replied:
I prefer it as it is, because pre-pg16 if there was something like FilledPermissionInfo it wouldn't have just requiredPerms.
The code may not be very beautiful right now, but there will be other PermissionInfo changes in the bigger PG16 PR, and this stays in line with them.

tejeswarm commented:
Thinking out load, shall we have one generic (new) file for all the routines that changed in Pg 16 i.e. all the new `..._compat(....) `routines from this PR? It will be easy for review and ideally, we should nuke this file when we stop supporting Pg15.

naisila commented:
@onderkalaci 
> We normally try to avoid merging into main thing that are not yet ready. Are you planning to fix those items before merging?

Yes, I plan to fix those before merging.

naisila commented:
@tejeswarm 
> Thinking out load, shall we have one generic (new) file for all the routines that changed in Pg 16 i.e. all the new `..._compat(....)` routines from this PR? It will be easy for review and ideally, we should nuke this file when we stop supporting Pg15.

We are using `pg_version_compat` file for that. Given that we only support 3 major versions the file is not huge, so I don't think we need to introduce another one.

naisila commented:
Update:
I still have to address the commits labelled with "FIX THIS". Other than that, it looks good.
@tejeswarm thanks for your reviews, I addressed your comments, and I cleaned up the PR. Please write comments with suggestions instead of pushing commits.
@onderkalaci still have to refactor the deadlock code a bit.

tejeswarm commented:
> Thinking out load, shall we have one generic (new) file for all the routines that changed in Pg 16 i.e. all the new `..._compat(....) `routines from this PR? It will be easy for review and ideally, we should nuke this file when we stop supporting Pg15.


Yes, I noticed all the macros there, which is good, I meant for functions, is it good to have a ".c" file?

tejeswarm commented:
> > Thinking out load, shall we have one generic (new) file for all the routines that changed in Pg 16 i.e. all the new `..._compat(....) `routines from this PR? It will be easy for review and ideally, we should nuke this file when we stop supporting Pg15.
> 
> Yes, I noticed all the macros there, which is good, I meant for functions, is it good to have a ".c" file?

Hmm, I just noticed, the routines are moved to .h file, I'm fine with this approach too.

tejeswarm commented on src/backend/distributed/planner/fast_path_router_planner.c:
```diff
@@ -103,15 +103,23 @@ PlannedStmt *
+	SeqScan *scanNode = makeNode(SeqScan);
+	Plan *plan = &(scanNode->scan.plan);
+#else
 	Scan *scanNode = makeNode(Scan);
```
Can we make this as `SeqScan` too?

In pre-16, both are same.

`typedef Scan SeqScan;`

tejeswarm replied:
```
#define get_plan_from_scan_node(node) node->scan.plan
#define get_plan_from_scan_node(node) node.plan
```

naisila replied:
Turns out SeqScan is already defined in all PG versions. It will get less readable to find a solution so I will just leave it as is.

tejeswarm commented on src/backend/distributed/planner/fast_path_router_planner.c:
```diff
@@ -103,15 +103,23 @@ PlannedStmt *
 	/* there is only a single relation rte */
+#if PG_VERSION_NUM >= PG_VERSION_16
+	scanNode->scan.scanrelid = 1;
+#else
```
#define     set_relid_in_scan_node(node, id)    node->scan.scanrelid = id
#define     set_relid_in_scan_node(node, id)    node->scanrelid = id

tejeswarm commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -1091,12 +1099,38 @@ columnar_vacuum_rel(Relation rel, VacuumParams *params,
+	Assert(TransactionIdPrecedesOrEquals(cutoffs.FreezeLimit, cutoffs.OldestXmin));
+
+	/*
+	 * Columnar storage doesn't hold any transaction IDs, so we can always
```
Does this logic apply to pre-16 versions too?

naisila replied:
Yes

tejeswarm commented on src/backend/columnar/columnar_customscan.c:
```diff
@@ -535,7 +541,11 @@ ColumnarIndexScanAdditionalCost(PlannerInfo *root, RelOptInfo *rel,
 	 * same stripe again and again.
 	 */
+#if PG_VERSION_NUM >= PG_VERSION_16
+	double absIndexCorrelation = fabs(indexCorrelation);
```
Can we have something like this in compat.h file?

```
#if PG_VERSION_NUM >= PG_VERSION_16
             #define float_abs(a) fabs(a)
#else 
             #define  float_abs(a) abs(a) 
```

and change the below to

`double absIndexCorrelation = float_abs(indexCorrelation);`

tejeswarm commented:
@naisila is it too early to turn on pg-16 tests for CI here?

tejeswarm commented:
> @naisila is it too early to turn on pg-16 tests for CI here?

Please ignore, I see you have another branch running tests

naisila commented:
> @naisila is it too early to turn on pg-16 tests for CI here?

@tejeswarm Since this branch only takes care of compilation, I have only turned on the build-16 test.

naisila commented:
[Try child permission info](https://github.com/citusdata/citus/pull/7005/commits/e896f7ede391b2522a5f8202779921472bcf6afc)

Tested locally by printing number of columns read in the columnar tests, both in main branch(with PG15) and in PG16 support branch, seems to be working as the printed number is the same in both tests in case the table is a child partition.

I believe the code can be simplified though, there should be an easier way to reach parent table's permission info.

naisila commented:
Update:
> I was able to figure out the issue. I am making use of append_rel_list in order to obtain the range table entry of the parent. However, apparently, there can be more that one (parent_id, child_id) pair in that list. I am just looking at the first one and breaking. Hence, it only sees the number 3. If I would continue, it would also see the number 1, which is the correct one in this case. So, I should fix this

It turns out that `RelOptInfo` in the `AddColumnarScanPath` function keeps track of the index of the child table, and I can directly check that. So, all tests are successful right now. However, I am now doing some extra testing for multi-level partitioning/inheritance.

naisila commented:
Update:
After thorough checking, all subcases can be covered by looking into the permission info of `reloptinfo->top_parent`. Will update the PR soon.

naisila commented:
We can leave the grant option commit as is to merge this PR. I am asking for another review.

onderkalaci reviewed (approved):
I don't see much value waiting this PR. Looks good to me

naisila commented on src/backend/distributed/planner/query_colocation_checker.c:
```diff
@@ -126,7 +126,7 @@ static RangeTblEntry *
 AnchorRte(Query *subquery)
 {
 	FromExpr *joinTree = subquery->jointree;
```
Yes, because we were passing FALSE for include_joins before.

naisila commented on src/backend/distributed/utils/relation_utils.c:
```diff
@@ -28,3 +31,37 @@ RelationGetNamespaceName(Relation relation)
+ * Given data consists of relid, inh and requiredPerms
+ * Took a quick look around Postgres, unless specified otherwise,
+ * we are dealing with GetUserId().
+ * Currently the following entries are filled like this:
```
I'd rather leave it as is, in order to not miss anything. There will be much more added for PermissionInfo in the bigger PG16 PR.
